### PR TITLE
Improve Model type definitions

### PR DESCRIPTION
**Pull request by bajtos**
- Define two new types: ModelSettings and ModelProperties to describe the objects describing model properties and settings.
- Add static property `base` and static method `extend` to the definition
of `ModelBase` class.
- Add property `defaultModelBaseClass` to `ModelBuilder` class definition.
- Fix `PropertyDefinition` interface: remove `name`, add optional `id`.

These issues were discovered while working on https://github.com/strongloop/loopback-next/pull/2274.

### Checklist

- New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: types/model.d.ts [hunk 1]
```diff
@@ -19,17 +19,17 @@ export type PropertyType =
  * Property definition
  */
 export interface PropertyDefinition extends AnyObject {
-  name: string;
```
Comment (bajtos) [inline types/model.d.ts]: AFAICT, the property name is always stored as the key in property map (name => definition).

File: types/model.d.ts [hunk 2]
```diff
@@ -19,17 +19,17 @@ export type PropertyType =
...
   type?: PropertyType;
+  id?: boolean;
```
Comment (raymondfeng) [inline types/model.d.ts]: In LB3, `id` is a `boolean` or `number`. See https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/model-definition.js#L132

Comment (bajtos) [reply types/model.d.ts]: Good catch! Fixed.

**Comment by bajtos** [comment]
@raymondfeng LGTY now?

**Pull request merged**